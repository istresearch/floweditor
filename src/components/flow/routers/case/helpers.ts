import { isRelativeDate } from 'components/flow/routers/helpers';
import { Operator, Operators } from 'config/interfaces';
import { getOperatorConfig } from 'config/operatorConfigs';
import {
  LessThan,
  MoreThan,
  Numeric,
  NumOrExp,
  Required,
  validate,
  IsValidIntent
} from 'store/validators';
import { titleCase } from 'utils';

import { CaseElementProps, CaseElementState } from './CaseElement';
import { SelectOption } from 'components/form/select/SelectElement';
import { Asset } from 'store/flowContext';
import i18n from 'config/i18n';

export const initializeForm = (props: CaseElementProps): CaseElementState => {
  const arg1 =
    props.kase.arguments && props.kase.arguments.length >= 1 ? props.kase.arguments[0] : '';
  const arg2 =
    props.kase.arguments && props.kase.arguments.length === 2 ? props.kase.arguments[1] : '';

  return {
    errors: [],
    operatorConfig: getOperatorConfig(props.kase.type),
    argument: { value: arg1 },
    min: { value: arg1 },
    max: { value: arg2 },
    state: { value: arg1 },
    district: { value: arg2 },
    intent: { value: arg1 ? { name: arg1, value: arg1 } : null },
    confidence: { value: arg2 },
    categoryName: { value: props.categoryName || '' },
    categoryNameEdited: !!props.categoryName,
    valid: true
  };
};
/**
 * Determines prefix for case's exit name
 */
export const prefix = (operatorType: string): string => {
  let pre = '';

  if (operatorType.indexOf('_lt') > -1) {
    if (operatorType.indexOf('date') > -1) {
      pre = 'Before ';
    } else {
      if (operatorType.indexOf('lte') > -1) {
        pre = '<= ';
      } else {
        pre = '< ';
      }
    }
  } else if (operatorType.indexOf('_gt') > -1) {
    if (operatorType.indexOf('date') > -1) {
      pre = 'After ';
    } else {
      if (operatorType.indexOf('gte') > -1) {
        pre = '>= ';
      } else {
        pre = '>';
      }
    }
  }

  return pre;
};

/**
 * Returns min, max values for Operators.has_number_between case
 */
export const getMinMax = (args: string[] = []): { min: string; max: string } => {
  let min = '';
  let max = '';
  if (args.length) {
    if (strContainsNum(args[0])) {
      min = args[0];
    }
    if (args[1]) {
      if (strContainsNum(args[1])) {
        max = args[1];
      }
    }
  }
  return {
    min,
    max
  };
};

export const isFloat = (val: string): boolean => /^[+-]?\d?(\.\d*)?$/.test(val.trim());

export const isInt = (val: string): boolean => /^[+-]?\d+$/.test(val.trim());

export const strContainsNum = (str: string): boolean => {
  const trimmed = str.trim();
  if (isFloat(trimmed)) {
    return true;
  } else if (isInt(trimmed)) {
    return true;
  } else {
    return false;
  }
};

export const parseNum = (str: string): number => {
  const trimmed = (str || '').trim();
  if (isFloat(trimmed)) {
    return parseFloat(str);
  } else if (isInt(trimmed)) {
    return parseInt(trimmed, 10);
  }
};

export const validateCase = (keys: {
  operatorConfig: Operator;
  argument?: string;
  state?: string;
  district?: string;
  min?: string;
  max?: string;
  confidence?: string;
  intent?: SelectOption;
  exitName?: string;
  exitEdited?: boolean;
  classifier?: Asset;
}): Partial<CaseElementState> => {
  // when the exit is set, our arguments become required
  const validators = keys.exitEdited && keys.exitName ? [Required] : [];

  const updates: Partial<CaseElementState> = {
    operatorConfig: keys.operatorConfig
  };

  updates.district = { value: '', validationFailures: [] };
  updates.state = { value: '', validationFailures: [] };
  updates.min = { value: '', validationFailures: [] };
  updates.max = { value: '', validationFailures: [] };
  updates.argument = { value: '', validationFailures: [] };
  updates.intent = { value: null, validationFailures: [] };
  updates.confidence = { value: '', validationFailures: [] };

  if (keys.operatorConfig.operands > 0) {
    switch (keys.operatorConfig.type) {
      case Operators.has_number_eq:
      case Operators.has_number_gt:
      case Operators.has_number_gte:
      case Operators.has_number_lt:
      case Operators.has_number_lte:
        validators.push(NumOrExp);
        break;
      case Operators.has_date_eq:
      case Operators.has_date_lt:
      case Operators.has_date_gt:
        validators.push(Numeric);
        break;
    }

    if (keys.operatorConfig.type === Operators.has_number_between) {
      const max = keys.max || '';
      const min = keys.min || '';

      const maxExpression = max.indexOf('@') > -1;
      const minExpression = min.indexOf('@') > -1;
      const hasExpression = maxExpression || minExpression;

      const numeric = [Numeric];

      updates.min = validate(
        i18n.t('forms.minimum_value', 'Minimum value'),
        min,
        validators
          .concat(!minExpression ? numeric : [])
          .concat(
            !hasExpression
              ? [LessThan(parseFloat(keys.max), i18n.t('forms.the_maximum', 'the maximum'))]
              : []
          )
      );

      updates.max = validate(
        i18n.t('forms.maximum_value', 'Maximum value'),
        max,
        validators
          .concat(!maxExpression ? numeric : [])
          .concat(
            !hasExpression
              ? [MoreThan(parseFloat(keys.min), i18n.t('forms.the_minimum', 'the minimum'))]
              : []
          )
      );
    } else if (keys.operatorConfig.type === Operators.has_district) {
      updates.argument = validate(
        i18n.t('forms.state', 'State'),
        keys.argument || '',
        validators.concat([])
      );
    } else if (keys.operatorConfig.type === Operators.has_ward) {
      updates.state = validate(
        i18n.t('forms.state', 'State'),
        keys.state || '',
        validators.concat([])
      );
      updates.district = validate(
        i18n.t('forms.district', 'District'),
        keys.district || '',
        validators.concat([])
      );
    } else if (
      keys.operatorConfig.type === Operators.has_top_intent ||
      keys.operatorConfig.type === Operators.has_intent
    ) {
      const intentValidators = [IsValidIntent(keys.classifier)];
      if (keys.confidence) {
        intentValidators.push(Required);
      }
      updates.intent = validate(i18n.t('forms.intent', 'Intent'), keys.intent, intentValidators);
      updates.confidence = validate(
        i18n.t('forms.confidence', 'Confidence'),
        keys.confidence || '',
        validators.concat(keys.intent ? [Numeric, Required] : [Numeric])
      );
    } else {
      updates.argument = validate('Value', keys.argument || '', validators);
    }
  }

  updates.categoryNameEdited = !!keys.exitEdited;
  updates.categoryName = validate(
    i18n.t('forms.category', 'Category'),
    updates.categoryNameEdited ? keys.exitName : getCategoryName(updates),
    updates.argument.value ||
      (updates.min.value && updates.max.value) ||
      (updates.state.value && updates.district.value)
      ? [Required]
      : []
  );

  updates.valid =
    updates.state.validationFailures.length === 0 &&
    updates.district.validationFailures.length === 0 &&
    updates.min.validationFailures.length === 0 &&
    updates.max.validationFailures.length === 0 &&
    updates.argument.validationFailures.length === 0 &&
    updates.intent.validationFailures.length === 0 &&
    updates.confidence.validationFailures.length === 0 &&
    updates.categoryName.validationFailures.length === 0;

  return updates;
};

export const getCategoryName = (state: Partial<CaseElementState>): string => {
  if (state.categoryNameEdited) {
    return state.categoryName.value;
  }

  if (state.operatorConfig.operands === 0) {
    return state.operatorConfig.categoryName;
  }

  if (
    state.operatorConfig.type === Operators.has_intent ||
    state.operatorConfig.type === Operators.has_top_intent
  ) {
    if (state.intent.value) {
      return titleCase(state.intent.value.name.replace('_', ' '));
    }
  }

  if (
    state.operatorConfig.type === Operators.has_number_between &&
    state.min.value &&
    state.max.value
  ) {
    return `${state.min.value} - ${state.max.value}`;
  }

  if (isRelativeDate(state.operatorConfig.type)) {
    const count = parseInt(state.argument.value, 10);
    if (!isNaN(count)) {
      const today =
        state.operatorConfig.type === Operators.has_date_eq
          ? i18n.t('forms.today_proper', 'Today')
          : i18n.t('forms.today', 'today');
      const op = count < 0 ? ' - ' : ' + ';
      const inDays =
        ' ' + (Math.abs(count) === 1 ? i18n.t('forms.day', 'day') : i18n.t('forms.days', 'days'));
      return prefix(state.operatorConfig.type) + today + op + Math.abs(count) + inDays;
    }
  }

  if (state.argument && state.argument.value) {
    const pre = prefix(state.operatorConfig.type);
    const words = state.argument.value.match(/\w+/g);

    if (words && words.length > 0) {
      const [firstWord] = words;
      return pre + titleCase(firstWord);
    }

    return pre + titleCase(state.argument.value);
  }

  return '';
};
