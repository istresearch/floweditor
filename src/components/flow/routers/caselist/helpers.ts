import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { Operator } from 'config/interfaces';
import { createUUID } from 'utils';

export const createEmptyCase = (operator: Operator): CaseProps => {
  const uuid = createUUID();
  return {
    uuid,
    kase: {
      uuid,
      type: operator.type,
      arguments: [''],
      category_uuid: null
    },
    categoryName: '',
    valid: true
  };
};
