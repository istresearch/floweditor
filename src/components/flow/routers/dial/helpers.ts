import { DialRouterFormState } from 'components/flow/routers/dial/DialRouterForm';
import { createRenderNode, getSwitchRouter } from 'components/flow/routers/helpers';
import { DIAL_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import {
  Case,
  Category,
  DialCategoryNames,
  DialStatus,
  Exit,
  Router,
  RouterTypes,
  SwitchRouter,
  Wait,
  WaitTypes
} from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { createUUID } from 'utils';

export const nodeToState = (settings: NodeEditorSettings): DialRouterFormState => {
  let phone = '';
  let resultName: StringEntry = { value: '' };

  if (settings.originalNode && getType(settings.originalNode) === Types.wait_for_dial) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      phone = router.wait.phone;
      resultName = { value: router.result_name || '' };
    }
  }

  return {
    phone: { value: phone },
    resultName,
    valid: true
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: DialRouterFormState
): RenderNode => {
  let categories: Category[] = [];
  let cases: Case[] = [];
  let exits: Exit[] = [];

  const previousRouter = settings.originalNode ? getSwitchRouter(settings.originalNode.node) : null;

  // see if we are editing an existing dial router so we reuse exits
  if (previousRouter && previousRouter.wait && previousRouter.wait.type === WaitTypes.dial) {
    previousRouter.cases.forEach(kase => cases.push(kase));
    previousRouter.categories.forEach(category => categories.push(category));
    settings.originalNode.node.exits.forEach((exit: any) => exits.push(exit));
  } else {
    // otherwise, let's create some new ones
    exits = [
      {
        uuid: createUUID(),
        destination_uuid: null
      },
      {
        uuid: createUUID(),
        destination_uuid: null
      },
      {
        uuid: createUUID(),
        destination_uuid: null
      },
      {
        uuid: createUUID(),
        destination_uuid: null
      }
    ];

    categories = [
      {
        uuid: createUUID(),
        name: DialCategoryNames.Answered,
        exit_uuid: exits[0].uuid
      },
      {
        uuid: createUUID(),
        name: DialCategoryNames.NoAnswer,
        exit_uuid: exits[1].uuid
      },
      {
        uuid: createUUID(),
        name: DialCategoryNames.Busy,
        exit_uuid: exits[2].uuid
      },
      {
        uuid: createUUID(),
        name: DialCategoryNames.Failure,
        exit_uuid: exits[3].uuid
      }
    ];

    cases = [
      {
        uuid: createUUID(),
        type: Operators.has_only_text,
        arguments: [DialStatus.answered],
        category_uuid: categories[0].uuid
      },
      {
        uuid: createUUID(),
        type: Operators.has_only_text,
        arguments: [DialStatus.noAnswer],
        category_uuid: categories[1].uuid
      },
      {
        uuid: createUUID(),
        type: Operators.has_only_text,
        arguments: [DialStatus.busy],
        category_uuid: categories[2].uuid
      }
    ];
  }

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const wait = { type: WaitTypes.dial, phone: state.phone.value } as Wait;

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    wait: wait,
    default_category_uuid: categories[categories.length - 1].uuid,
    categories,
    cases,
    operand: DIAL_OPERAND,
    ...optionalRouter
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.wait_for_dial,
    []
  );

  return newRenderNode;
};
