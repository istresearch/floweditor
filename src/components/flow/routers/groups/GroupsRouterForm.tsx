import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { RouterFormProps } from 'components/flow/props';
import { GROUP_LABEL } from 'components/flow/routers/constants';
import { nodeToState, stateToNode } from 'components/flow/routers/groups/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import TypeList from 'components/nodeeditor/TypeList';
import { fakePropType } from 'config/ConfigProvider';
import { Asset } from 'store/flowContext';
import { AssetArrayEntry, FormState, mergeForm, StringEntry } from 'store/nodeEditor';
import { Alphanumeric, Required, StartIsNonNumeric, validate } from 'store/validators';
import i18n from 'config/i18n';
import { renderIssues } from 'components/flow/actions/helpers';

// TODO: Remove use of Function
// tslint:disable:ban-types
export interface GroupsRouterFormState extends FormState {
  groups: AssetArrayEntry;
  resultName: StringEntry;
}

export default class GroupsRouterForm extends React.Component<
  RouterFormProps,
  GroupsRouterFormState
> {
  public static contextTypes = {
    endpoints: fakePropType,
    assetService: fakePropType
  };

  constructor(props: RouterFormProps) {
    super(props);
    this.state = nodeToState(this.props.nodeSettings);

    bindCallbacks(this, {
      include: [/^handle/]
    });
  }

  private handleGroupsChanged(groups: Asset[]): void {
    this.handleUpdate({ groups });
  }

  private handleUpdateResultName(resultName: string): void {
    this.handleUpdate({ resultName });
  }

  private handleUpdate(keys: { groups?: Asset[]; resultName?: string }): boolean {
    const updates: Partial<GroupsRouterFormState> = {};

    if (keys.hasOwnProperty('groups')) {
      updates.groups = validate(i18n.t('forms.groups', 'Groups'), keys.groups, [Required]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        Alphanumeric,
        StartIsNonNumeric
      ]);
    }

    const updated = mergeForm(this.state, updates);
    this.setState(updated);
    return updated.valid;
  }

  private handleSave(): void {
    if (this.state.valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p>{GROUP_LABEL}</p>
        <AssetSelector
          name={i18n.t('forms.groups', 'Groups')}
          assets={this.props.assetStore.groups}
          entry={this.state.groups}
          searchable={true}
          onChange={this.handleGroupsChanged}
          multi={true}
        />
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }
}
