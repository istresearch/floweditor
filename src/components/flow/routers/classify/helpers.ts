import {
  hasCases,
  createCaseProps,
  resolveRoutes,
  ResolvedRoutes,
  createRenderNode
} from 'components/flow/routers/helpers';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Types, Operators, VISIBILITY_HIDDEN } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { CallClassifier, SwitchRouter, Case, Exit, Category, RouterTypes } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry, FormEntry } from 'store/nodeEditor';
import { createUUID, scalarArrayEquals, snakify } from 'utils';
import { ClassifyRouterFormState } from 'components/flow/routers/classify/ClassifyRouterForm';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { getOperatorConfig } from 'config';

export const getOriginalAction = (settings: NodeEditorSettings): CallClassifier => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 && settings.originalNode.node.actions[0]);

  if (action.type === Types.call_classifier) {
    return action as CallClassifier;
  }
};

export const nodeToState = (settings: NodeEditorSettings): ClassifyRouterFormState => {
  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: 'Result' };
  let initialCases: CaseProps[] = [];

  let operand = DEFAULT_OPERAND;
  let classifier: FormEntry = { value: null };

  let hiddenCases: CaseProps[] = [];

  if (getType(settings.originalNode) === Types.split_by_intent) {
    const router = settings.originalNode.node.router as SwitchRouter;

    if (hasCases(settings.originalNode.node)) {
      initialCases = createCaseProps(router.cases, settings.originalNode);

      hiddenCases = initialCases.filter(
        (kase: CaseProps) => getOperatorConfig(kase.kase.type).visibility === VISIBILITY_HIDDEN
      );

      initialCases = initialCases.filter(
        (kase: CaseProps) => getOperatorConfig(kase.kase.type).visibility !== VISIBILITY_HIDDEN
      );
    }

    resultName = { value: router.result_name || 'Result' };
    const action = getOriginalAction(settings) as CallClassifier;
    classifier = { value: action.classifier };
    operand = action.input;
  }

  const state: ClassifyRouterFormState = {
    hiddenCases,
    resultName,
    classifier,
    operand: { value: operand },
    cases: initialCases,
    valid: true
  };

  return state;
};

export interface Route {
  type: Operators;
  arguments: string[];
  name: 'Other';
}

export const ensureRoute = (routes: ResolvedRoutes, route: Route) => {
  const existingCasePosition = routes.cases.findIndex(
    kase => kase.type === route.type && scalarArrayEquals(kase.arguments, route.arguments)
  );

  // if it already exists, make sure it's at the end
  if (existingCasePosition >= 0) {
    // already in the right place we are done
    if (existingCasePosition === routes.cases.length - 1) {
      return;
    }

    const [existingCase] = routes.cases.splice(existingCasePosition, 1);
    routes.cases.push(existingCase);

    return;
  }

  const exit: Exit = {
    uuid: createUUID()
  };

  const category: Category = {
    uuid: createUUID(),
    name: route.name,
    exit_uuid: exit.uuid
  };

  // otherwise let's add it
  const kase: Case = {
    uuid: createUUID(),
    type: route.type,
    arguments: route.arguments,
    category_uuid: category.uuid
  };

  routes.categories.push(category);
  routes.cases.push(kase);

  // our new exit should go right in front of the default exit
  routes.exits.splice(routes.exits.length - 1, 0, exit);
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ClassifyRouterFormState
): RenderNode => {
  let uuid = createUUID();

  const routes = resolveRoutes(
    [...state.cases, ...state.hiddenCases],
    false,
    settings.originalNode.node,
    'Failure'
  );

  // make sure we have an other route since failure is our default
  ensureRoute(routes, {
    type: Operators.has_category,
    arguments: ['Success', 'Skipped'],
    name: 'Other'
  });

  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const routerResultName = state.resultName.value;
  const actionResultName = '_' + routerResultName + ' Classification';

  const newAction: CallClassifier = {
    uuid,
    type: Types.call_classifier,
    result_name: actionResultName,
    input: state.operand.value,
    classifier: {
      uuid: state.classifier.value.uuid,
      name: state.classifier.value.name
    }
  };

  const router: SwitchRouter = {
    cases: routes.cases,
    operand: `@results.${snakify(actionResultName)}`,
    categories: routes.categories,
    type: RouterTypes.switch,
    default_category_uuid: routes.defaultCategory,
    result_name: routerResultName
  };

  return createRenderNode(
    settings.originalNode.node.uuid,
    router,
    routes.exits,
    Types.split_by_intent,
    [newAction]
  );
};
