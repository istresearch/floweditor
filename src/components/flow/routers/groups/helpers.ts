import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { GroupsRouterFormState } from 'components/flow/routers/groups/GroupsRouterForm';
import { createRenderNode, getSwitchRouter, resolveRoutes } from 'components/flow/routers/helpers';
import { GROUPS_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Category, FlowNode, RouterTypes, SwitchRouter } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { createUUID } from 'utils';

export const nodeToState = (settings: NodeEditorSettings): GroupsRouterFormState => {
  const state: GroupsRouterFormState = {
    groups: { value: [] },
    resultName: { value: '' },
    valid: false
  };

  if (getType(settings.originalNode) === Types.split_by_groups) {
    state.groups.value = extractGroups(settings.originalNode.node);
    state.resultName = {
      value: (settings.originalNode.node.router as SwitchRouter).result_name || ''
    };
    state.valid = true;
  }

  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: GroupsRouterFormState
): RenderNode => {
  const currentCases = groupsToCases(state.groups.value);
  const { cases, exits, defaultCategory: defaultExit, caseConfig, categories } = resolveRoutes(
    currentCases,
    false,
    settings.originalNode.node
  );

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    cases,
    categories,
    default_category_uuid: defaultExit,
    operand: GROUPS_OPERAND,
    result_name: state.resultName.value
  };

  return createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.split_by_groups,
    [],
    { cases: caseConfig }
  );
};

export const extractGroups = (node: FlowNode): any[] => {
  let groups: any[] = [];
  const router = getSwitchRouter(node);
  if (router) {
    groups = (router as SwitchRouter).cases.map(kase => {
      const category = router.categories.find((cat: Category) => cat.uuid === kase.category_uuid);
      return {
        name: category.name,
        uuid: kase.arguments[0]
      };
    });
  }
  return groups;
};
export const groupsToCases = (groups: any[] = []): CaseProps[] => {
  return groups.map(({ name, uuid }: any) => ({
    uuid,
    kase: {
      uuid: createUUID(),
      type: Operators.has_group,
      arguments: [uuid, name],
      category_uuid: ''
    },
    categoryName: name,
    valid: true
  }));
};
