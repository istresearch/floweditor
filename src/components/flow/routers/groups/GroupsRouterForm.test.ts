import GroupsRouterForm from 'components/flow/routers/groups/GroupsRouterForm';
import { extractGroups } from 'components/flow/routers/groups/helpers';
import { SwitchRouter } from 'flowTypes';
import { composeComponentTestUtils, mock, setMock } from 'testUtils';
import {
  createGroupsRouterNode,
  getRouterFormProps,
  SubscribersGroup
} from 'testUtils/assetCreators';
import * as utils from 'utils';

const groupsRouterNode = createGroupsRouterNode();
const { setup } = composeComponentTestUtils(GroupsRouterForm, getRouterFormProps(groupsRouterNode));

mock(utils, 'createUUID', utils.seededUUIDs());

describe(GroupsRouterForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should update and save', () => {
      const { instance, props } = setup(true, { updateRouter: setMock() });
      instance.handleGroupsChanged([SubscribersGroup]);
      instance.handleUpdateResultName('My Group Result');
      expect(instance.state).toMatchSnapshot();

      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel changes', () => {
      const { instance, props } = setup(true, { updateRouter: setMock() });
      instance.handleGroupsChanged([SubscribersGroup]);
      instance.handleUpdateResultName('My Group Result');
      instance.getButtons().secondary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });

    it('validates before saving', () => {
      const { instance, props } = setup(true, { updateRouter: setMock() });
      instance.handleGroupsChanged([]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });
  });

  describe('helpers', () => {
    describe('extractGroups', () => {
      it('should extract groups from the exits of a groupsRouter node', () => {
        extractGroups(groupsRouterNode.node).forEach((group, idx) => {
          expect(group.name).toBe(groupsRouterNode.node.router.categories[idx].name);
          expect(group.uuid).toBe(
            (groupsRouterNode.node.router as SwitchRouter).cases[idx].arguments[0]
          );
          expect(group).toMatchSnapshot();
        });
      });
    });
  });
});
