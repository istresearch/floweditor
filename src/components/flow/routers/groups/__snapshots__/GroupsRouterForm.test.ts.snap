// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupsRouterForm helpers extractGroups should extract groups from the exits of a groupsRouter node 1`] = `
Object {
  "name": "Customers",
  "uuid": "23ff7152-b588-43e4-90de-fda77aeaf7c0",
}
`;

exports[`GroupsRouterForm helpers extractGroups should extract groups from the exits of a groupsRouter node 2`] = `
Object {
  "name": "Unsatisfied Customers",
  "uuid": "2429d573-80d7-47f8-879f-f2ba442a1bfd",
}
`;

exports[`GroupsRouterForm helpers extractGroups should extract groups from the exits of a groupsRouter node 3`] = `
Object {
  "name": "Early Adopters",
  "uuid": "cdbf9e01-aaa7-4381-8259-ee042447bcac",
}
`;

exports[`GroupsRouterForm helpers extractGroups should extract groups from the exits of a groupsRouter node 4`] = `
Object {
  "name": "Testers",
  "uuid": "afaba971-8943-4dd8-860b-3561ed4f1fe1",
}
`;

exports[`GroupsRouterForm helpers extractGroups should extract groups from the exits of a groupsRouter node 5`] = `
Object {
  "name": "Subscribers",
  "uuid": "33b28bac-b588-43e4-90de-fda77aeaf7c0",
}
`;

exports[`GroupsRouterForm helpers extractGroups should extract groups from the exits of a groupsRouter node 6`] = `
Object {
  "name": "Champions",
  "uuid": "e0cccbcf-49be-4eef-88a7-ce2150a6d902",
}
`;

exports[`GroupsRouterForm render should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_groups"
  title="Split by Group Membership"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by group membership",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split by Group Membership",
        "type": "split_by_groups",
      }
    }
    onChange={[MockFunction]}
  />
  <p>
    Select the groups you'd like to split by below
  </p>
  <AssetSelector
    entry={
      Object {
        "value": Array [
          Object {
            "name": "Customers",
            "uuid": "23ff7152-b588-43e4-90de-fda77aeaf7c0",
          },
          Object {
            "name": "Unsatisfied Customers",
            "uuid": "2429d573-80d7-47f8-879f-f2ba442a1bfd",
          },
          Object {
            "name": "Early Adopters",
            "uuid": "cdbf9e01-aaa7-4381-8259-ee042447bcac",
          },
          Object {
            "name": "Testers",
            "uuid": "afaba971-8943-4dd8-860b-3561ed4f1fe1",
          },
          Object {
            "name": "Subscribers",
            "uuid": "33b28bac-b588-43e4-90de-fda77aeaf7c0",
          },
          Object {
            "name": "Champions",
            "uuid": "e0cccbcf-49be-4eef-88a7-ce2150a6d902",
          },
        ],
      }
    }
    multi={true}
    name="Groups"
    onChange={[Function]}
    searchable={true}
  />
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`GroupsRouterForm updates should update and save 1`] = `
Object {
  "groups": Object {
    "validationFailures": Array [],
    "value": Array [
      Object {
        "name": "Subscriber",
        "uuid": "68223118-109f-442a-aed3-7bb3e1eab687",
      },
    ],
  },
  "resultName": Object {
    "validationFailures": Array [],
    "value": "My Group Result",
  },
  "valid": true,
}
`;

exports[`GroupsRouterForm updates should update and save 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "68223118-109f-442a-aed3-7bb3e1eab687",
              "Subscriber",
            ],
            "category_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "type": "has_group",
            "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Subscriber",
            "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
          },
          Object {
            "exit_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
            "name": "Other",
            "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
          },
        ],
        "default_category_uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        "operand": "@contact.groups",
        "result_name": "My Group Result",
        "type": "switch",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
    },
    "ui": Object {
      "config": Object {
        "cases": Object {},
      },
      "position": null,
      "type": "split_by_groups",
    },
  },
]
`;
